Flesh out README & add API docs

barrucadu opened the pull request:
The README is shorter than our Account Manager or Attribute Service apps, eg it doesn't have detailed set-up documentation.

Our GOV.UK READMEs have been standardising on a fairly terse format[(1)][][(2)][][(3)][] which I've copied.  In particular, there are no specific GOV.UK Docker commands referenced, I asked why on Slack and the thinking is that it makes GOV.UK Docker seem more magical than it is, and makes people think that they have to use GOV.UK Docker to run our apps, when really it's nothing more than a convenient shortcut.

The API documentation is inspired by [publishing-api][] and [router-api][].

[(1)]: https://github.com/alphagov/imminence
[(2)]: https://github.com/alphagov/router-api
[(3)]: https://github.com/alphagov/search-api
[publishing-api]: https://github.com/alphagov/publishing-api/blob/master/docs/api.md
[router-api]: https://github.com/alphagov/router-api/blob/master/docs/api-endpoints.md

---

[Trello card](https://trello.com/c/MBJ9DNmU/678-write-some-documentation-for-the-account-api)

huwd commented on README.md:
```diff
@@ -1,6 +1,85 @@
+[tc-save-results]: http://finder-frontend.dev.gov.uk/transition-check/save-your-results?c%5B%5D=living-ie
+[tc-start]: http://finder-frontend.dev.gov.uk/transition-check/questions
+
+
```
Nitpick, double new line?

barrucadu replied:
I normally put a double new line before a h2, though I see I didn't do that for the `Further documentation` heading.

huwd reviewed (approved):
Content looks good and clear.
Nice one!

Would be interested to hear what @jonathanglassman thinks?

jonathanglassman commented:
> Content looks good and clear.
> Nice one!
> 
> Would be interested to hear what @jonathanglassman thinks?

The API endpoints content looks good. I'm a bit confused about the Requirements for API consumers section - would be good to have a chat with @barrucadu about it.

jonathanglassman commented on docs/api.md:
```diff
@@ -0,0 +1,248 @@
+# API documentation
+
```
Recommend having an introduction to the API in terms of what the user is trying to achieve. For example:

This API is for GOV.UK services that want to add user sign in / sign out and attribute storage functionality to their applications. When you have integrated this API with your service, you can sign in and authenticate your users' identities on GOV.UK.

barrucadu replied:
Added

> This API is for GOV.UK Publishing microservices (the things which make
> up www.gov.uk) to implement personalisation and user session
> management.  It is not for other government services.

jonathanglassman commented on docs/api.md:
```diff
@@ -0,0 +1,248 @@
+  - [`POST /api/transition-checker-email-subscription`](#post-apitransition-checker-email-subscription)
+- [Healthcheck](#healthcheck)
+
+## Nomenclature
```
Can we use a more plain English word, for example, definitions or defined terms?

barrucadu replied:
Maybe, but "nomenclature" is what other GOV.UK projects are using, eg:

- https://github.com/alphagov/email-alert-api#nomenclature
- https://github.com/alphagov/content-publisher/#nomenclature
- https://github.com/alphagov/support-api/#nomenclature
- https://github.com/alphagov/finder-frontend/#nomenclature

barrucadu replied:
I'll see if I can rework the rest of the doc to not need custom terminology

jonathanglassman commented on docs/api.md:
```diff
@@ -0,0 +1,248 @@
+- `GOVUK-Account-Session`: sent to our CDN to update the user's session cookie.
+- `GOVUK-Account-End-Session`: sent to our CDN to delete the user's session cookie.
+
+Every app which uses this API is responsible for using the custom
```
Recommend making more active and focused on the service. For example:

"If your service's app uses this API, you are responsible for..."

jonathanglassman commented on docs/api.md:
```diff
@@ -0,0 +1,248 @@
+The [gds-api-adapters][] methods for these endpoints have a required
+`govuk_account_session:` parameter, which must be set to the latest
+Session Identifier.
+
```
It looks to me like I, the developer on the service team who wants to integrate this service with the API, need to do two things here:

- set the `GOVUK-Account-Session` request header
- set the required `govuk_account_session:` parameter to the latest session identifier

Is that correct? Does setting one set the other?

Also, how do I set them? What do I actually need to do, and does this documentation need to tell me how to do this? @huwd suggested putting in an example request here.

jonathanglassman commented on docs/api.md:
```diff
@@ -0,0 +1,248 @@
+
+If an Access Restricted endpoint returns a `401: Unauthenticated`
+response, then the user must no longer be considered logged in, and
+the end-session response header must be set when serving the response
```
passive => active

jonathanglassman reviewed (changes requested):
API reference looks good. For other sections, have suggested clarifications and asked questions where necessary.

barrucadu commented:
Thanks for the comments.  @jonathanglassman sorry for the confusion, I don't think I mentioned this anywhere, but this API is specifically for GOV.UK internal use, like [publishing-api](https://github.com/alphagov/publishing-api/), [router-api](https://github.com/alphagov/router-api/), or [email-alert-api](https://github.com/alphagov/email-alert-api/) - it's not something which other service teams will come into contact with.

barrucadu commented:
I've removed the "Access Restricted" term (editing other sections where needed), reworked the "Requirements for API consumers" section, added an "Example API usage" section, and added missing descriptions to the two `transition-checker-email-subscription` endpoints.

jonathanglassman commented on docs/api.md:
````diff
@@ -0,0 +1,339 @@
+```
+
+
+## API endpoints
````
@barrucadu apologies I forgot to include some feedback on this section. Best practice for API reference documentation is to include the following:

- Endpoints and methods
- Parameters and arguments
- Example requests and responses for each endpoint with example parameters
- Error response codes

This content has the first two. Is it possible to include the example requests and responses, and error response codes? Happy to discuss at any point.

barrucadu replied:
:+1: I've added response codes and example requests & responses.

barrucadu commented:
I've reworked the doc a fair bit so I'll resolve the outdated comments.
